### DISCUSSION TITLE
BrainVISA crashes on launch

### DISCUSSION THREAD

**Discussion by HdeV999** [brainvisa/brainvisa.github.io#130] [Q&A] [answered]
Hello, 
I've been facing an issue since a few days which I haven't managed to solve by myself: 
while my bv and anatomist commands work normally, my brainvisa crashes on launch, with the error message:
[14528:14528:0217/124726.007770:FATAL:zygote_host_impl_linux.cc(165)] Check failed: process.IsValid(). Failed to launch zygote process
I've attached the entire error thread at the end of the message just in case. I believe I messed something up with my environment while trying to install new python libraries to my environment, but this was a few weeks ago so I'm a bit hazy on what I did which could have impacted my environment. I think I upgraded and downgraded my PyQt5 install at some point. 
I've tried reinstalling a whole new brainvisa from the website but the bug was carried over: interestingly enough, without any databases, the initial "update database" window opened, and brainvisa crashed after this step. 
Do you have any clues on what can be going wrong? On how I could clean my installation or eventually clean reinstall it? I don't think my local environment is affecting this (I may be wrong though), because I'm working from a cluster and my colleagues using the same cluster are not facing this issue...
Best wishes,

Full error thread:
[14528:14528:0217/124726.007770:FATAL:zygote_host_impl_linux.cc(165)] Check failed: process.IsValid(). Failed to launch zygote process
#0 0x2b0f2e659aee <unknown>
#1 0x2b0f2e66c0e2 <unknown>
#2 0x2b0f2e276a4d <unknown>
#3 0x2b0f2e275b56 <unknown>
#4 0x2b0f2e2760fe <unknown>
#5 0x2b0f2df21d55 <unknown>
#6 0x2b0f2df25262 <unknown>
#7 0x2b0f2dd13a9b <unknown>
#8 0x2b0f2dd14bd5 <unknown>
#9 0x2b0f2dd03bdd <unknown>
#10 0x2b0f2dd03f0e QtWebEngineCore::WebContentsAdapter::WebContentsAdapter()
#11 0x2b0f2d3cd091 <unknown>
#12 0x2b0f2d3cdc04 QWebEnginePage::QWebEnginePage()
#13 0x2b0f2d18ae39 sipQWebEnginePage::sipQWebEnginePage()
#14 0x2b0f2d18afe2 init_type_QWebEnginePage
#15 0x2b0dc6aea204 <unknown>
#16 0x0000004b4c96 <unknown>
#17 0x0000005a00ce PyObject_Call
#18 0x00000050d617 _PyEval_EvalFrameDefault
#19 0x000000507f94 <unknown>
#20 0x000000509272 _PyFunction_FastCallDict
#21 0x000000595351 <unknown>
#22 0x00000054a39f <unknown>
#23 0x000000551f81 <unknown>
#24 0x0000005aa2ac _PyObject_FastCallKeywords
#25 0x00000050a7f3 <unknown>
#26 0x00000050c274 _PyEval_EvalFrameDefault
#27 0x000000507f94 <unknown>
#28 0x000000509cc0 <unknown>
#29 0x00000050a6bd <unknown>
#30 0x00000050c274 _PyEval_EvalFrameDefault
#31 0x000000507f94 <unknown>
#32 0x0000005092e7 _PyFunction_FastCallDict
#33 0x000000595351 <unknown>
#34 0x00000054a39f <unknown>
#35 0x000000551f81 <unknown>
#36 0x0000005aa2ac _PyObject_FastCallKeywords
#37 0x00000050a7f3 <unknown>
#38 0x00000050c274 _PyEval_EvalFrameDefault
#39 0x000000507f94 <unknown>
#40 0x0000005092e7 _PyFunction_FastCallDict
#41 0x000000595351 <unknown>
#42 0x00000054a39f <unknown>
#43 0x000000551f81 <unknown>
#44 0x0000005aa2ac _PyObject_FastCallKeywords
#45 0x00000050a7f3 <unknown>
#46 0x00000050c274 _PyEval_EvalFrameDefault
#47 0x000000509085 _PyFunction_FastCallDict
#48 0x000000595351 <unknown>
#49 0x00000054a39f <unknown>
#50 0x000000551f81 <unknown>
#51 0x0000005aa2ac _PyObject_FastCallKeywords
#52 0x00000050a7f3 <unknown>
#53 0x00000050c274 _PyEval_EvalFrameDefault
#54 0x000000509989 <unknown>
#55 0x00000050a6bd <unknown>
#56 0x00000050c274 _PyEval_EvalFrameDefault
#57 0x000000509989 <unknown>
#58 0x00000050a6bd <unknown>
#59 0x00000050c274 _PyEval_EvalFrameDefault
#60 0x000000507f94 <unknown>
#61 0x000000516585 <unknown>

Fatal Python error: Aborted

Thread 0x00002b0fd9026700 (most recent call first):
  File "/usr/lib/python3.6/threading.py", line 299 in wait
  File "/usr/lib/python3.6/threading.py", line 551 in wait
  File "/casa/host/build/python/soma_workflow/engine.py", line 558 in start_loop
  File "/casa/host/build/python/soma_workflow/engine.py", line 102 in run
  File "/usr/lib/python3.6/threading.py", line 916 in _bootstrap_inner
  File "/usr/lib/python3.6/threading.py", line 884 in _bootstrap

Thread 0x00002b0fd8e25700 (most recent call first):
  File "/usr/lib/python3.6/threading.py", line 299 in wait
  File "/usr/lib/python3.6/threading.py", line 551 in wait
  File "/casa/host/build/python/soma_workflow/schedulers/local_scheduler.py", line 291 in _iterate
  File "/casa/host/build/python/soma_workflow/schedulers/local_scheduler.py", line 137 in loop
  File "/usr/lib/python3.6/threading.py", line 864 in run
  File "/usr/lib/python3.6/threading.py", line 916 in _bootstrap_inner
  File "/usr/lib/python3.6/threading.py", line 884 in _bootstrap

Thread 0x00002b0fd8c24700 (most recent call first):
  File "/casa/host/build/python/soma_workflow/schedulers/local_scheduler.py", line 205 in poll_processes
  File "/usr/lib/python3.6/threading.py", line 864 in run
  File "/usr/lib/python3.6/threading.py", line 916 in _bootstrap_inner
  File "/usr/lib/python3.6/threading.py", line 884 in _bootstrap

Current thread 0x00002b0dc13b1f40 (most recent call first):
  File "/casa/host/build/python/soma/qt4gui/text.py", line 192 in __init__
  File "/casa/host/build/python/soma/qt4gui/text.py", line 243 in __init__
  File "/casa/host/build/python/brainvisa/processing/qt4gui/neuroProcessesGUI.py", line 1007 in __init__
  File "/casa/host/build/python/brainvisa/processing/qt4gui/neuroProcessesGUI.py", line 1160 in __init__
  File "/casa/host/build/python/brainvisa/processing/qt4gui/neuroProcessesGUI.py", line 4124 in __init__
  File "/casa/host/build/python/brainvisa/processing/qt4gui/neuroProcessesGUI.py", line 4876 in showMainWindow
  File "/casa/host/build/brainvisa/neuro.py", line 175 in main
  File "/casa/host/build/brainvisa/neuro.py", line 216 in <module>
  File "<string>", line 9 in execfile
  File "<string>", line 33 in <module>
Aborted
INFO:    Cleanup error: while unmounting /var/apptainer/mnt/session/underlay/lib directory: device or resource busy, while unmounting /var/apptainer/mnt/session/final directory: device or resource busy

**Comment by sapetnioc** [comment]
Hi HdeV,

It's good to read you. I hope you are doing well.

From the error message, the error is very likely related to Qt : zygote term seems to be used by Chromium that underlies the Qt widget to display HTML content. There is no simpler way to make our software crash than changing PyQt. I believe that Anatomist still work because it did not try to load the Qt version from PyQt but our compiled version. If you try to use PyQt from Anatomist you will probably have the same kind of funny (because unpredictable) result.

The reason for that is that PyQt from Pypi comes with its own Qt. Even if it can have compatible version we provide our own Qt because we do not know any way to compile C++ software using the Qt provided by PyQt. And some of our software is in C++. PyQt's Qt and BrainVISA's Qt are not binary compatible but, to tell it simply, all C++ software that uses Qt must use the same Qt binary. And PyQt is such a software. Since BrainVISA may use some Python modules that are in C++ and are using Qt, everything becomes linked with this binary dependency. So, to use BrainVISA, one must use our Qt and our PyQt.

I think you don't have to rebuild your environment. You probably have modified PyQt via `pip` from within the container. Therefore packages had been installed in your BrainVISA environment home directory. It may be possible to uninstall them by `pip`. Or you can remove the whole home Python directory (something like `<brainvisa_dir>/home/.local/lib/python3.10`) and reinstall the ones you need (except PyQt).

Yours sincerely.

**Comment by sapetnioc** [comment]
Sorry, I missed that you still got the error after reinstalling BrainVISA. My solution is probably wrong then. Did you modify the Singularity image or are you using the same home directory (for the two installs) ? We may get in touch directly to find out how to solve this, it would be more efficient.

**Reply by HdeV999** [reply]
Yes that sounds great. Would you have any time soon for a small videoconference maybe? I just sent you an email in case you don't have my lasting email address...

**Comment by sapetnioc** [comment]
We checked that your installation is clean. There is no custom pip installed package therefore I was not looking towards a good solution. Since it works for someone else in the same condition and after reading the errors again, I suspect something related to a low level system problem. It seems that there are problems with Python threading (both in Qt and in Soma-Workflow). Could it be possible that Python cannot properly run a thread anymore in your environment (you told me that you removed some cluster modules) ? I told you to wait but I changed my mind, it may be a good idea to get in touch with your support team either to ask them if they have an idea or simply to reset your environment.

**Comment by sapetnioc** [comment]
Did you removed your comment ? I received a mail but cannot see it her. Anyway, you can test threading with the following Python code. 
```
from __future__ import print_function

import threading
import time

def thread_function(t):
print('here')
time.sleep(t)
print('there')


t = threading.Thread(target=thread_function, args=(2,))
t.start()
t.join()
```

**Reply by HdeV999** [reply]
Yes I did, I felt bad about asking you this question when I could easily find my answer online...
So I just tested this code both within and without the bv bash environment and it worked fine, printing here and there 2 seconds apart. 

Apart from that I reinstalled a new casa-distro and brainvisa image (casa-dev-5.3-17) and when opening brainvisa, I first get the upload/update databases window and then brainvisa crashes, with a much less verbose error message: 

connect to computing resource: login-e-12
Trace/breakpoint trap
[120467:120467:0100/000000.104343:ERROR:zygote_linux.cc(646)] write: Broken pipe (32)

Still zygote being unhappy...

**Comment by HdeV999** [comment]
I have useful new information: I have just assisted a new colleague in installing the new 5.1.0 release of brainvisa. Everything (bv, anatomist) went well until we tried to open brainvisa: she got this error message: 

connect to computing resource: login-e-15
Trace/breakpoint trap
[217289:217289:0100/000000.051516:ERROR:zygote_linux.cc(646)] write: Broken pipe (32)
Traceback (most recent call last):
  File "/casa/casa-distro/python/casa_distro/configuration_gui.py", line 536, in close_and_launch
    subprocess.check_call(command)
  File "/usr/lib/python3.10/subprocess.py", line 369, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'brainvisa' returned non-zero exit status 133.
Aborted

So what I have now is: 
- a colleague with a brainvisa developer installation dating from November - works well
- a colleague with a fresh new classic brainvisa installation from a fresh new environment - brainvisa crashes with a zygote_linux error
- a new developer install with a linux_zygote error
- an old developer install with a linux_zygote error (on which I have ran bv_maker recently)
I therefore currently believe that the modification that happened that broke my setup may be related to a bv_maker: my colleague with a working install did not bv_make in the last few months. 
Any insights on that?

**Comment by denisri** [comment, marked as answer]
Hi,
I don't know what is zygote, but is seems to be in the Qt web browser module: QtWebEngine. This module replaces the older QtWebKit, and has been activated somewhat by mistake in the 5.1.0 release. So we can try using the older one again. This will likely fix the problem for you, but is not a long run solution because Qt6 will not provide QtWebKit any longer. Anyway for now it should fix things. Note that I have no idea why it doesn't work on your system in a container, whereas it works well for us in the same container.
Since you have a developer setup, just run `bv_maker` to update and start `brainvisa` again. Let's see if this works before going further...

**Reply by HdeV999** [reply]
That worked! My brainvisa is up and running again \o/
Thank you so much

**Reply by denisri** [reply]
Then we have to provide a fix for the binary distribution...
In singularity-based binary distribs, the brainvisa installation is read-only since it is inside the read-only singularity image.
When you run the `bv` command (without arguments) you get a configuration GUI. There you can make a read-write installation (which will, of course eat a bit less than 3 Gb of additional disk space). Then you can get updated files from https://brainvisa.info/download/updates then take the version of brainvisa (5.1.0 here) and you may find fixed files at the location they have in the installed distribution.
A read-write install is found, from the container side, in `/casa/host/install/`. On the host filesystem, it is in `<environment_dir>/install/` for a user install.
I am adding in the `bv` GUI a tool to install all updates in one click, but it will be available in the next release only.

**Reply by HdeV999** [reply]
Ok thanks!

**Comment by denisri** [comment]
I have opened a ticket here: brainvisa/axon#66
